devito/passes/clusters/aliases.py

```diff
@@ -88,6 +144,101 @@ def collect(exprs):
     return aliases
 
 
+def extract(exprs, aliases):
+    """
+    Extract the candidate aliases.
+    """
+    is_time_invariant = make_is_time_invariant(exprs)
+    time_invariants = {e.rhs: is_time_invariant(e) for e in exprs}
+
+    processed = []
+    candidates = OrderedDict()
+    for e in exprs:
+        # Cost check (to keep the memory footprint under control)
+        naliases = len(aliases.get(e.rhs))
+        cost = estimate_cost(e, True)*naliases
+        test0 = lambda: cost >= MIN_COST_ALIAS and naliases > 1
+        test1 = lambda: cost >= MIN_COST_ALIAS_INV and time_invariants[e.rhs]
+        if test0() or test1():
+            candidates[e.rhs] = e.lhs
+        else:
+            processed.append(e)
+
+    return candidates, processed
+
+
+def process(candidates, aliases, cluster, template):
+    """
+    Create Clusters from aliasing expressions.
+    """
+    clusters = []
+    subs = {}
+    for origin, alias in aliases.items():
+        if all(i not in candidates for i in alias.aliased):
+            continue
+
+        # The write-to Intervals
+        writeto = [Interval(i.dim, *alias.relaxed_diameter.get(i.dim, (0, 0)))
+                   for i in cluster.ispace.intervals if not i.dim.is_Time]
+        writeto = IntervalGroup(writeto)
+
+        # Optimization: no need to retain a SpaceDimension if it does not
+        # induce a flow/anti dependence (below, `i.offsets` captures this, by
+        # telling how much halo will be needed to honour such dependences)
+        dep_inducing = [i for i in writeto if any(i.offsets)]
+        try:
+            index = writeto.index(dep_inducing[0])
+            writeto = IntervalGroup(writeto[index:])
+        except IndexError:
+            perf_adv("Couldn't optimize some of the detected redundancies")
```

Comment:
Super pedantic I know, but "Could not optimize ..." reads better in such situations.